### PR TITLE
Merge registered and latest and move to apimachinery

### PR DESCRIPTION
**Pull request by nikhiljindal**
Fixes https://github.com/kubernetes/kubernetes/issues/19275

Ref https://github.com/kubernetes/kubernetes/issues/19273

Merges registered and latest as discussed in https://github.com/kubernetes/kubernetes/issues/19275.
Also moves them to a new apimachinery pkg as discussed in https://github.com/kubernetes/kubernetes/issues/19273.

Apart from `pkg/apimachinery/registered/registered.go`, all other changes are simply find replace `latest`->`registered` and `pkg/api/registered`->`pkg/apimachinery/registered`

@kubernetes/sig-api-machinery @lavalamp @caesarxuchao @smarterclayton

### CODE REVIEW

**Comment by nikhiljindal** [comment]
@smarterclayton Looks like it will take a while for github.com/GoogleCloudPlatform/kubernetes/pull/17922 to merge.

This is mostly just a rename PR. Let me know if you still want this to be blocked on github.com/GoogleCloudPlatform/kubernetes/pull/17922

File: pkg/apimachinery/registered/registered.go [hunk 1]
```diff
@@ -59,15 +69,27 @@ func init() {
...
+	if _, found := groupMetaMap[groupName]; found {
+		return fmt.Errorf("group %v is already registered", groupMetaMap)
+	}
+	groupMetaMap[groupName] = &groupMeta
+	return nil
+}
+
+// EnableVersions adds the versions for the given group to the list of enabled versions.
+// Note that the caller should call RegisterGroup before calling this method.
+// The caller of this
```
Comment (caesarxuchao) [inline pkg/apimachinery/registered/registered.go]: remove the new line

File: pkg/apimachinery/registered/registered.go [hunk 2]
```diff
@@ -43,6 +47,12 @@ var (
 	envRequestedVersions = map[unversioned.GroupVersion]struct{}{}
 )
 
+// ExternalVersions is a list of all external versions for the legacy API
+// group in order of most preferred to least preferred
+var LegacyGroupVersions = []unversioned.GroupVersion{
```
Comment (caesarxuchao) [inline pkg/apimachinery/registered/registered.go]: It this still needed? Can we use `EnabledVersionsForGroup()` instead?

File: pkg/apimachinery/registered/registered.go [hunk 3]
```diff
@@ -99,12 +121,6 @@ func IsEnabledVersion(v unversioned.GroupVersion) bool {
 	return found
 }
 
-// IsRegisteredVersion returns if a version is registered.
-func IsRegisteredVersion(v unversioned.GroupVersion) bool {
```
Comment (caesarxuchao) [inline pkg/apimachinery/registered/registered.go]: Why do you remove this function?

File: pkg/apimachinery/registered/registered.go [hunk 4]
```diff
@@ -59,15 +69,27 @@ func init() {
...
+func RegisterVersions(availableVersions []unversioned.GroupVersion) {
+	for _, v := range availableVersions {
 		registeredVersions[v] = struct{}{}
 	}
 }
 
-// EnableVersions add the versions to the enabledVersions. The caller of this
-// function is responsible to add the version to 'latest' and 'Scheme'.
+// RegisterGroup adds the given group to the list of registered groups.
+func RegisterGroup(groupMeta apimachinery.GroupMeta) error {
```
Comment (caesarxuchao) [inline pkg/apimachinery/registered/registered.go]: How about moving all the functions that are moved from latest.go to their own file, say "group_meta.go"? The existing code in this file is not focused on the concept of group, while these functions are, so we can naturally split them in two files.

**Comment by caesarxuchao** [comment]
I took a quick pass, made a few suggestions, generally SGTM.

File: pkg/apimachinery/registered/registered.go (hunk 3 above)
Comment (nikhiljindal) [inline pkg/apimachinery/registered/registered.go]: Its dead code (no one is using it).

File: pkg/apimachinery/registered/registered.go (hunk 2 above)
Comment (nikhiljindal) [inline pkg/apimachinery/registered/registered.go]: Sure. Will remove this and use EnabledVersionsForGroup(api.GroupName) instead

File: pkg/apimachinery/registered/registered.go (hunk 3 above)
Comment (caesarxuchao) [inline pkg/apimachinery/registered/registered.go]: I remember @deads2k specifically asked me to add these functions when reviewing my PR. @deads2k do you still want to keep these?

File: pkg/apimachinery/registered/registered.go (hunk 4 above)
Comment (nikhiljindal) [inline pkg/apimachinery/registered/registered.go]: I am planning to merge RegisterGroup and RegisterVersion into one.
GroupMeta has all the required information.

It requires moving enableVersion logic as well (since we cannot create GroupMeta and RestMapper unless objects have been added to schema), which I did not do in this PR.
Will send another PR for it.

File: pkg/apimachinery/registered/registered.go (hunk 1 above)
Comment (nikhiljindal) [inline pkg/apimachinery/registered/registered.go]: Done

File: pkg/apimachinery/registered/registered.go (hunk 2 above)
Comment (nikhiljindal) [inline pkg/apimachinery/registered/registered.go]: Done

**Comment by nikhiljindal** [comment]
Thanks @caesarxuchao!
Updated code as per comments.
PTAL

File: pkg/apimachinery/registered/registered.go (hunk 3 above)
Comment (deads2k) [inline pkg/apimachinery/registered/registered.go]: > I remember @deads2k specifically asked me to add these functions when reviewing my PR. @deads2k do you still want to keep these?

Yes, having a specific method like this makes deconflicting requested bits from available bits in something like the API Installer.  I haven't gotten a chance to go through the code where its created in the master, but I wanted them to resolve item 2 in this list: https://github.com/kubernetes/kubernetes/issues/16173

Comment (nikhiljindal) [inline pkg/apimachinery/registered/registered.go]: Yes right now availabeVersions, enabledVersions and groupMeta have overlapping information.
I plan to refactor that in another PR (this PR is just code moving, no logic change).
We will see what methods are required after the refactor and add any thats missing.

No point in keeping dead code around.

File: pkg/apimachinery/doc.go [hunk 5]
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2014 The Kubernetes Authors All rights reserved.
```
Comment (lavalamp) [inline pkg/apimachinery/doc.go]: Nit: I think we don't generally change the year for moves...

**Comment by lavalamp** [comment]
This is a big improvement!

File: pkg/apimachinery/doc.go (hunk 5 above)
Comment (nikhiljindal) [inline pkg/apimachinery/doc.go]: Fixed

**Comment by nikhiljindal** [comment]
Thanks @lavalamp 
Pushed a new version as per comments.

Can I have the LGTM label?

**Comment by caesarxuchao** [comment]
@smarterclayton do you still want this PR to wait for #17922? I think you'll get several conflicts at the use of `lastest.Group(...).Codec`

**Comment by smarterclayton** [comment]
It's fine for this to go first, I think @lavalamp will take a while to sign off and I don't mind rebasing.

**Comment by lavalamp** [comment]
This is pretty mechanical-- LGTM

I'll try to look at #17922 soon.

**Comment by nikhiljindal** [comment]
Rebased. Hopefully this time the e2e finishes before I have to rebase again :)
